# SPNtypeID

SPNtypeID is a [NextFlow](https://www.nextflow.io/) pipeline used for assembly of SPN whole genome sequence data and identification of serotype.

### Table of Contents:
[Usage](#using-the-pipeline)  
[Workflow outline](#workflow-outline)  
[Read trimming and quality assessment](#read-trimming-and-quality-assessment)  
[Genome assembly](#genome-assembly)  
[Assembly quality assessment](#assembly-quality-assessment)  
[Genome coverage](#genome-coverage)  
[Contamination detection](#contamination-detection)
[Serotyping](#serotyping)                                                                                                                                   
[Output](#output-files)  

### Using the workflow
The pipeline is designed to start from raw Illumina reads. All reads must be in the same directory. Then start the pipeline using:
```
nextflow SPNtypeID.nf --reads [path-to-reads] -profile [docker,singularity,aws]
```

You can specify a version of the pipeline and run it directly from the github repository by using:
```
nextflow wslh-bio/SPNtypeID -r [version] --reads [path-to-reads] -profile [docker,singularity,aws]
```

### Workflow outline

#### Read trimming and quality assessment
Read trimming and cleaning is performed using [BBtools v38.76](https://jgi.doe.gov/data-and-tools/bbtools/) to trim reads of low quality bases and remove PhiX contamination. Then [FastQC v0.11.8](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/) is used assess the quality of the raw and cleaned reads.

#### Genome assembly
Assembly of the cleaned and trimmed reads is performed using [Shovill v1.1.0](https://github.com/tseemann/shovill).

#### Assembly quality assessment
Quality assessment of the assemblies is performed using [QUAST v5.0.2](http://bioinf.spbau.ru/quast)

#### Genome coverage
Mean and median genome coverage is determined by mapping the cleaned reads back their the assembly using [BWA v0.7.17-r1188](http://bio-bwa.sourceforge.net/) and calculating depth using [samtools v1.10](http://www.htslib.org/)

#### Contamination detection
Contamination is detected by classifying reads using [Kraken v1.0.0](https://ccb.jhu.edu/software/kraken2/).

#### Serotpying
Serotyping is performed using [SeroBA v1.0.2](https://github.com/sanger-pathogens/seroba).

### Output files

```
spntypeid_results
├── assembled
│   └── *.contigs.fa
├── cg_pipeline
│   ├── logs
│   │   └── *.error.log (optional)
│   ├── *.qual.tsv
│   └── quality_results.tsv
├── fastqc
│   ├── fastqc_summary.tsv
│   ├── *.html
│   └── zips
│       └── *.zip
├── kraken
│   ├── kraken_results.tsv
│   ├── *.kraken2.txt
│   └── logs
│       └── *.error.log (optional)
├── mapping
│   ├── bams
│   │   ├── *.bam
│   │   └── *.bai
│   ├── coverage_stats.tsv
│   ├── depth
│   │   └── *.depth.tsv
│   ├── sams
│   │   └── *.sam
│   └── stats
│       └── *.stats.txt
├── quast
│   ├── quast_results.tsv
│   ├── *.quast.report.tsv
│   └── *.transposed.quast.report.tsv
├── multiqc_report.html
├── seroba
│   ├── logs
│   │   └── *.error.log (optional)
│   ├── *_detailed_serogroup_info.txt*
│   ├── *.pred.tsv
│   └── seroba_results.tsv
├── spntypeid_report.csv
└── trimming
    ├── bbduk_results.tsv
    ├── reads
    │   └── *_clean_*    
    └── stats
        └── *.trim.txt
```

**\*.contigs.fa** - Shovill assembly for each sample  
**\*.error.log** - Error logs for each sample from the cg pipeline step  
**\*.qual.tsv** - Report of quality results for each sample  
**\*quality_results** - Summary table of quality results  
**fastqc_summary.tsv** - Summary table of FastQC results  
**\*.html** - HTML files of FastQC results  
**\*.zip** - Zipped folders of FastQC output  
**kraken_results.tsv** - Summary table of Kraken results  
**\*.kraken2.txt** - Report of Kraken results for each sample  
**\*.error.log** - Error logs for each sample from the Kraken step  
**\*.bam** - Alignments to an assembly BAM format  
**\*.bai** - Index file of alignments to an assembly  
**coverage_stats.tsv** - Summary table of mean and median coverage calculated with Samtools depth  
**\*.depth.tsv** - Raw Samtools depth output for reads mapped to an assembly  
**\*.sam** - Alignments to an assembly SAM format  
**\*.stats.txt** - Output of Samtools stats for reads mapped to an assembly  
**quast_results.tsv** - Summary table of QUAST results  
**\*.quast.report.tsv** - Long-format QUAST results for each sample  
**\*.transposed.quast.report.tsv** - Wide-format QUAST results for each sample  
**multiqc_report.html** - HTML report generated by MultiQC  
**\*.error.log** - Error logs for each sample from the SeroBA step  
**\*\_detailed_serogroup_info.txt\*** - Detailed serogroup for each sample information if available  
**\*.pred.tsv** - SeroBA results for each sample  
**seroba_results.tsv** - Summary table of SeroBA results  
**spntypeid_report.csv** - Summary table of each step in SPNtypeID  
**bbduk_results.tsv** - Summary table of trimming with BBDuk  
**\*\_clean\_\*** - Trimmed and cleaned reads  
**\*.trim.txt** - Trimming results from BBDuk each sample  

### Authors
[Kelsey Florek](https://github.com/k-florek), WSLH Senior Genomics and Data Scientist  
[Abigail Shockey](https://github.com/AbigailShockey), WSLH Bioinformatician and Data Scientist
